### PR TITLE
Plumb CRDT KClass through storage to drivers.

### PR DESCRIPTION
**Pull request by jasonwyatt**
Add dataClass to the Store constructors, and pass it to the driver provider so drivers can be aware of the type of data they're tasked with storing.

### CODE REVIEW

File: java/arcs/core/storage/BackingStore.kt [hunk 1]
```diff
@@ -90,7 +92,17 @@ class BackingStore<Data : CrdtData, Op : CrdtOperation, T>(
...
         val store = DirectStore.CONSTRUCTOR(
             // Copy of our options, but with a child storage key using the muxId.
-            options.copy(options.storageKey.childKeyWithComponent(muxId))
+            options.copy(options.storageKey.childKeyWithComponent(muxId)),
+            dataClass = when (val type = options.type) {
+                is CrdtModelType<*, *, *> -> type.crdtModelDataClass
+                is ReferenceType<*> -> when (val contained = type.containedType) {
+                    is CrdtModelType<*, *, *> -> contained.crdtModelDataClass
+                    else -> throw UnsupportedOperationException(
+                        "Unsupported contained type: ${options.type}"
```
Comment (csilvestrini) [inline java/arcs/core/storage/BackingStore.kt]: It think this should be `$contained`

File: java/arcs/core/storage/BackingStore.kt [hunk 2]
```diff
@@ -90,7 +92,17 @@ class BackingStore<Data : CrdtData, Op : CrdtOperation, T>(
...
+                    )
+                }
+                else -> throw UnsupportedOperationException("Unsupported type: ${options.type}")
```
Comment (csilvestrini) [inline java/arcs/core/storage/BackingStore.kt]: `$type`

Comment (jasonwyatt) [reply java/arcs/core/storage/BackingStore.kt]: done

File: java/arcs/core/storage/BackingStore.kt (hunk 1 above)
Comment (jasonwyatt) [reply java/arcs/core/storage/BackingStore.kt]: done

**Pull request merged**